x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items_details.ts

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { QueryDslQueryContainer } from "@elastic/elasticsearch/lib/api/typesWithBodyKey";
+import { AGENT_NAME, AGENT_VERSION, SERVICE_ENVIRONMENT, SERVICE_LANGUAGE_NAME, SERVICE_NAME, SERVICE_NODE_NAME } from "@kbn/apm-plugin/common/elasticsearch_fieldnames";
+import { environmentQuery } from "@kbn/apm-plugin/common/utils/environment_query";
+import { AgentName } from "@kbn/apm-plugin/typings/es_schemas/ui/fields/agent";
+import { ProcessorEvent } from "@kbn/observability-plugin/common/processor_event";
+import { kqlQuery, rangeQuery } from "@kbn/observability-plugin/server/utils/queries";
+import { RandomSampler } from "../../lib/helpers/get_random_sampler";
+import { AgenItemsSetup } from "./get_agents_items";
+
+export function serviceNameQuery(
+  serviceName?: string
+): QueryDslQueryContainer[] {
+  if (!serviceName) {
+    return [];
+  }
+
+  return [{ term: { [SERVICE_NAME]: serviceName } }];
+}
+
+export function agentLanguageQuery(
+  agentName?: string
+): QueryDslQueryContainer[] {
+  if (!agentName) {
+    return [];
+  }
+
+  return [{ term: { [SERVICE_LANGUAGE_NAME]: agentName } }];
+}
+
+interface AggregationParams {
+  environment: string;
+  serviceName?: string;
+  agentLanguage?: string;
+  kuery: string;
+  setup: AgenItemsSetup;
+  maxNumServices: number;
+  start: number;
+  end: number;
+  randomSampler: RandomSampler;
+}
+
+export async function getAgentsDetails({
+  environment,
+  agentLanguage,
+  serviceName,
+  kuery,
+  setup,
+  maxNumServices,
+  start,
+  end,
+  randomSampler,
+}: AggregationParams) {
+  const { apmEventClient } = setup;
+
+  const response = await apmEventClient.search(
+    'get_agent_details',
+    {
+      apm: {
+        events: [ProcessorEvent.metric],
+      },
+      body: {
+        track_total_hits: false,
+        size: 0,
+        query: {
+          bool: {
+            filter: [
+              {
+                exists: {
+                  field: AGENT_NAME,
+                }
+              },
+              {
+                exists: {
+                  field: AGENT_VERSION,
+                }
+              },
+              {
+                exists: {
+                  field: SERVICE_NODE_NAME,
+                }
+              },
+              ...rangeQuery(start, end),
+              ...environmentQuery(environment),
+              ...kqlQuery(kuery),
+              ...serviceNameQuery(serviceName),
+              ...agentLanguageQuery(agentLanguage),
```

Comment:
[termQuery](https://github.com/elastic/kibana/blob/0ce440c44f0b72409282587bf2e1eda8f5142695/x-pack/plugins/observability/server/utils/queries.ts#L16) could be used for service name and agent language